src/realm/object-store/sync/sync_manager.cpp

```diff
@@ -452,6 +452,38 @@ void SyncManager::remove_user(const std::string& user_id)
     }
 }
 
+void SyncManager::delete_user(const std::string& user_id)
+{
+    util::CheckedLockGuard lock(m_user_mutex);
+    // Avoid itterating over m_users twice by not calling `get_user_for_identity`.
+    auto it = std::find_if(m_users.begin(), m_users.end(), [&user_id](auto& user) {
+        return user->identity() == user_id;
+    });
+    auto user = it == m_users.end() ? nullptr : *it;
+
+    if (!user)
+        return;
+
+    // Deletion should happen immediately, not when we do the cleanup
+    // task on next launch.
+    m_users.erase(it);
+    user->detach_from_sync_manager();
+
+    if (m_current_user && m_current_user->identity() == user->identity())
+        m_current_user = nullptr;
+
+    util::CheckedLockGuard fs_lock(m_file_system_mutex);
+    if (!m_metadata_manager)
+        return;
+
+    for (size_t i = 0; i < m_metadata_manager->all_unmarked_users().size(); i++) {
+        auto metadata = m_metadata_manager->all_unmarked_users().get(i);
+        if (user->identity() == metadata.identity()) {
+            metadata.remove();
```

Comment:
Before removing the entry from the metadata, this user's Realm files should also be deleted from disk, see https://github.com/realm/realm-core/blob/master/src/realm/object-store/sync/sync_manager.cpp#L139